src/DotNetLightning.Core/Crypto/Keys.fs

```diff
@@ -0,0 +1,660 @@
+        let perCommitmentPointBytes = this.ToBytes()
+        let tweak =
+            Key.FromHash <| Array.append perCommitmentPointBytes basepointBytes
+        PubKey.Add(basepoint, tweak.PubKey)
+
+    member this.DerivePaymentPrivKey (paymentBasepointSecret: PaymentBasepointSecret)
+                                         : PaymentPrivKey =
+        PaymentPrivKey <|
+            this.DerivePrivKey (paymentBasepointSecret.RawKey())
+
+    member this.DerivePaymentPubKey (paymentBasepoint: PaymentBasepoint)
+                                        : PaymentPubKey =
+        PaymentPubKey <|
+            this.DerivePubKey (paymentBasepoint.RawPubKey())
+
+    member this.DeriveDelayedPaymentPrivKey (delayedPaymentBasepointSecret: DelayedPaymentBasepointSecret)
+                                                : DelayedPaymentPrivKey =
+        DelayedPaymentPrivKey <|
+            this.DerivePrivKey (delayedPaymentBasepointSecret.RawKey())
+
+    member this.DeriveDelayedPaymentPubKey (delayedPaymentBasepoint: DelayedPaymentBasepoint)
+                                               : DelayedPaymentPubKey =
+        DelayedPaymentPubKey <|
+            this.DerivePubKey (delayedPaymentBasepoint.RawPubKey())
+
+    member this.DeriveHtlcPrivKey (htlcBasepointSecret: HtlcBasepointSecret)
+                                      : HtlcPrivKey =
+        HtlcPrivKey <|
+            this.DerivePrivKey (htlcBasepointSecret.RawKey())
+
+    member this.DeriveHtlcPubKey (htlcBasepoint: HtlcBasepoint)
+                                     : HtlcPubKey =
+        HtlcPubKey <|
+            this.DerivePubKey (htlcBasepoint.RawPubKey())
+
+    member this.DeriveRevocationPubKey (revocationBasepoint: RevocationBasepoint)
+                           : RevocationPubKey =
+        let revocationBasepointBytes = revocationBasepoint.ToBytes()
+        let perCommitmentPointBytes = this.ToBytes()
+        let revocationBasepointTweak =
+            Key.FromHash <| Array.append revocationBasepointBytes perCommitmentPointBytes
+        let perCommitmentPointTweak =
+            Key.FromHash <| Array.append perCommitmentPointBytes revocationBasepointBytes
+
+        RevocationPubKey <| PubKey.Add(
+            PubKey.Mul(revocationBasepoint.RawPubKey(), revocationBasepointTweak),
+            PubKey.Mul(this.RawPubKey(), perCommitmentPointTweak)
+        )
+
+    member this.DeriveCommitmentPubKeys (channelPubKeys: ChannelPubKeys)
+                                            : CommitmentPubKeys = {
+        RevocationPubKey =
+            this.DeriveRevocationPubKey channelPubKeys.RevocationBasepoint
+        PaymentPubKey =
+            this.DerivePaymentPubKey channelPubKeys.PaymentBasepoint
+        DelayedPaymentPubKey =
+            this.DeriveDelayedPaymentPubKey channelPubKeys.DelayedPaymentBasepoint
+        HtlcPubKey =
+            this.DeriveHtlcPubKey channelPubKeys.HtlcBasepoint
+    }
+
+type CommitmentNumber =
+    | CommitmentNumber of UInt48
```

Comment:
`Uint48` is a value type (and its inner value is value type too) so it is good to make this type struct too.

Reply:
Fixed.